### ISSUE TITLE
Issue with conda dependencies with custom channels

### ISSUE THREAD

**Issue by gregd33** [bentoml/BentoML#1429]
**Describe the bug**

This is related to some issues I've mentioned where you're building in an environment with _only_ access to specific repositories.

In short, the Dockerfile will call a script that does `conda env update -n base -f ./enviornment.yml`. The problem is that this will also use the global conda settings. Unlike `conda install` there is no `--override-channels` argument. 

So the build will break in such an environment

**To Reproduce**

This is quite difficult reproduce. Other than building in such an environment, you could create a service with a specific conda url as well the  `conda_overwrite_channels=True` which will create an `environment.yml` with only that url. If you add a `-v` to the `conda env update` you should see it installing from other urls. 

**Expected behavior**

It should be the case that _only_ the conda channels specified are used. 

**Potential Fix**

One fix is to add the line `conda config --system --remove channels defaults` in the `bentoml-init.sh` file, just before the`conda env update` line. This will remove the default urls so that the only channels used are specified in the environment.yml file. 

Since `conda` is not used after this step it should not cause other issues. There might be other ways - I've looked for a way to do this via the `conda env update` command with no luck. 

A safer approach might be to do this iff the `conda_overwrite_channels=True` but this would involve interacting with the `bentoml-init.sh` file when building which currently isn't done.

**Comment by bojiang** [comment]
In some cases, we need conda to install Python from the default channels.
Specifying the package with the full description `<channel_name>::<package_name>=<version>[build=<build>]` will do the trick. And this is always the recommended way to lock a package under conda's _smart_ dependency resolver.

**Comment by gregd33** [comment]
Just to clarify, the solution would be to specify the conda dependencies in the environment.yml (or I guess in the bento service which generates it) in that manner?

**Comment by parano** [comment]
@gregd33 yes I believe you can use that syntax directly when specifying conda dependencies in BentoML, e.g.:

```
@env(conda_dependencies=["chanelname::modulename=X.Y.Z"])
class MyMLSvc(BentoService):
    ...
```

**Comment by gregd33** [comment]
I don't think this works. Here is a way you can confirm yourself. Create an environment.yml file

channels:
- https://conda.anaconda.org/conda-forge/linux-64
dependencies:
- https://conda.anaconda.org/conda-forge/linux-64::mamba

And  then run `conda env update -n bento10 -f ./environment.yml  -vv` you'll see calls to the main anaconda repo. It seems to be looking for repodata.json

DEBUG urllib3.connectionpool:_new_conn(959): Starting new HTTPS connection (1):a repo.anaconda.com:443
DEBUG urllib3.connectionpool:_new_conn(959): Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG urllib3.connectionpool:_new_conn(959): Starting new HTTPS connection (1): repo.anaconda.com:443

I've also tested your suggestion on my system where anaconda is unavailable and the same happens.

**Comment by bojiang** [comment]
> It seems to be looking for repodata.json

Why is this a problem if the final version of `mamba` installed by conda is from channel `conda-forge/linux-64`?

If I am wrong, please let me know.
So the problem now is not to ensure the installation source channel of a certain conda package, but to use conda in an environment where `repo.anaconda.com` cannot be accessed, or to avoid access to `repo.anaconda.com` completely, right?

**Comment by gregd33** [comment]
>  but to use conda in an environment where repo.anaconda.com cannot be accessed

Yes that is exactly the use case. Sorry if that was unclear. 

For now, I have a script that basically does the fix above (removes default channels) and then does the docker build. So its workaround-able if nothing can be done.

**Comment by axelgobletbdr** [comment]
We are running into this same issue when building the bento images in our cluster. The generated `bentoml-init.sh` script contains the hardcoded `pkgs/main::python=$DESIRED_PY_VERSION` install. @bojiang you mentioned that

> In some cases, we need conda to install Python from the default channels.

What is the reason for this? Would it be possible to parameterize this setting? I think this would be beneficial to many enterprise users, as often direct access to the internet is blocked.

**Comment by yubozhao** [comment]
@axelgobletbdr We are using conda ensure the docker image has the python version (major.minor) that used for saving the bento bundles.

If you are using the generated Dockerfile, it has the current python version for the bento bundle. It won't use conda to install Python.

> Would it be possible to parameterize this setting? I think this would be beneficial to many enterprise users, as often direct access to the internet is blocked.

I think this is a good suggestion.  Can you create an issue for this?

**Comment by yubozhao** [comment]
cc @bojiang

**Comment by parano** [comment]
Discussed offline, the suggested fix is avoiding installing Python from conda, but require the user to ensure that the correct Python version was installed on their docker base image when the `custom_docker_image` option was used

**Comment by axelgobletbdr** [comment]
> avoiding installing Python from conda in the bentoml-init.sh, but require the user to ensure that the correct Python version was installed on their docker base image when the custom_docker_image option was used

I guess that this works. I still do not understand why the hardcoded channel is required in the image. Loosening that constraint would be another option if that has no undesired side effects.

**Comment by parano** [comment]
@axelgobletbdr here's a big more context about the proposed change: BentoML's default base image already have Python installed: https://github.com/bentoml/BentoML/blob/v0.11.0/docker/model-server/Dockerfile, That's why most of the time, BentoBundle's docker build process does not need to install python. In the `bentoml-init.sh` file, it only runs the `conda install pkgs/main::python=$DESIRED_PY_VERSION` line if the user is using their own docker image and their docker image does not have the correct python version installed. 

Not specifying the channel here can cause a few other issues, one being the default conda channel may not have the newer Python releases that the user might be using in their training environment(BentoML tries to keep the python version information and make sure it is consistent across model packaging and model serving).  We found it hard to automate this config for the user in the case of a custom docker base image, and it can be a lot cleaner to just ask the user to ensure the python version in their own docker image. Usually, only advanced users are using custom docker base images anyway, it shouldn't be a problem for these type of users.

**Comment by axelgobletbdr** [comment]
Okay, that is clear to me then. For us, shipping python with the base image would work. This could be enforced automatically when using a non-standard base image, or via an additional flag, which is up to you.

**Comment by bojiang** [comment]
@gregd33 Found this https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#sharing-an-environment
> You can exclude the default channels by adding nodefaults to the channels list... This is equivalent to passing the --override-channels option to most conda commands.

**Comment by bojiang** [comment]
> Since conda is not used after this step it should not cause other issues. There might be other ways - I've looked for a way to do this via the conda env update command with no luck.

I tested the way mentioned above from the conda official docs. Appending `nodefaults` to the end of the conda_channels works well.

```python
@env(conda_channels=["your.conda.repo", "nodefault"], conda_overwrite_channels=True)
```

It requires no changes of BentoML.

**Comment by parano** [comment]
@bojiang I find the current `conda_channels` and `conda_overwrite_channels` options to be very confusing. `conda_channels` essentially means "additional conda channels", I'd suggest making the `conda_channels` parameter override the default `["default"]` channel. In our docs it can say:

```
conda_channels: list of conda channels to use for installing conda packages specified with this environment, the default value is ["default"], containing only the default conda channel.
```

We will also need to update all our docs and example notebooks, where the `conda_channel` is used, e.g. [here](https://docs.bentoml.org/en/latest/concepts.html#conda-packages) we will need to change `[h2oai]` to `[default, h2oai]`

**Comment by parano** [comment]
> One fix is to add the line conda config --system --remove channels defaults in the bentoml-init.sh file, just before theconda env update line. This will remove the default urls so that the only channels used are specified in the environment.yml file.

Another approach is to run `conda config --system --remove channels defaults` in our docker base image build process, after we create the conda environment and install python
https://github.com/bentoml/BentoML/blob/master/docker/model-server/Dockerfile#L16

**Comment by gregd33** [comment]
> > Since conda is not used after this step it should not cause other issues. There might be other ways - I've looked for a way to do this via the conda env update command with no luck.
> 
> I tested the way mentioned above from the conda official docs. Appending `nodefaults` to the end of the conda_channels works well.
> 
> ```python
> @env(conda_channels=["your.conda.repo", "nodefault"], conda_overwrite_channels=True)
> ```
> 
> It requires no changes of BentoML.

Thanks, I will give that a try tomorrow! This is a very reasonable approach and should have been easier to find in the conda documentation.

**Comment by bojiang** [comment]
> I find the current conda_channels and conda_overwrite_channels options to be very confusing. conda_channels essentially means "additional conda channels", I'd suggest making the conda_channels parameter override the default ["default"] channel.

According to conda's official docs, to restore an environment from an `env.yml` file, conda already sees `channels` as the additional and preferred channels. I suggest making the behavior consistent with that rather.

That is, 
1. applying `conda_channels` as the `channels` (override not append)
2. when `conda_overwrite_channels=True`, append a `nodefaults` to the channels.

**Comment by parano** [comment]
@bojiang that sounds great!

**Comment by bojiang** [comment]
@parano Good. I'll make the PR accordingly.

**Comment by gregd33** [comment]
Just wanted to confirm that the `nodefaults` entry in the conda fix and the fix in https://github.com/bentoml/BentoML/pull/1414 combined will make a build work when anaconda.com is not available.

**Comment by parano** [comment]
@gregd33 that is correct, after #1501, the user will just need to specify ```@env(conda_override_channels=True)```, the `nodefaults` entry will be added to avoid using the "defaults" channel for any user specified conda dependencies

**Comment by gregd33** [comment]
Miscommunication on my part - I was confirming that `nodefaults` did, in fact, work for me in such an environment (with a temporary fix for #1414) after testing. Thanks  for the work on this issue!
